R/pseudocolor_plot.R

```diff
@@ -0,0 +1,346 @@
+#' y <- rnorm(1000)
+#'
+#' get_density(x, y, n = 100)
+#'
+.get2Ddensity <- function(
+    x,
+    y,
+    n = 500,
+    ...
+) {
+  # Check that MASS is installed
+  expect_MASS()
+
+  dens_grid <-
+    MASS::kde2d(x, y, n = n, ...)
+
+  dens <-
+    dens_grid$z[cbind(findInterval(x, dens_grid$x),
+                      findInterval(y, dens_grid$y))]
+
+  return(dens)
+
+}
+
+#' Create a pseudocolor plot.
+#'
+#' Create a pseudocolor plot of two markers.
+#'
+#' @param object A Seurat object.
+#' @param marker1 Marker to plot along the x-axis.
+#' @param marker2 Marker to plot along the y-axis.
+#' @param facet_vars Variables to facet the plot by.
+#' @param plot_gate A data.frame with columns 'xmin', 'xmax', 'ymin', 'ymax' to plot a gate. This data.frame can also
+#'                  contain the variables in 'facet_vars' to plot different gates by facets.
+#' @param scale_density Scale the density to the maximum density per facet.
+#' @param margin_density Add marginal density plots. Only supported when 'facet_vars' is NULL.
+#' @param coord.fixed Fix the aspect ratio of the plot. Only supported when 'margin_density' is FALSE.
+#' @param pt.size Size of the points.
+#' @param alpha Transparency of the points.
+#' @param layer Name of layer to plot.
+#' @param grid_n Number of grid points to calculate the density.
+#' @param ... Additional arguments to pass to 'MASS::kde2d'.
+#'
+#' @return A ggplot object.
+#'
+#' @import rlang
+#'
+#' @examples
+#'
+#' library(pixelatorR)
+#' library(Seurat)
+#'
+#' # A mock-up Seurat Object
+#' object <-
+#'  CreateSeuratObject(counts = matrix(c(rpois(100000, 40),
+#'                                       rpois(100000, 5))[sample(1:200000, 200000)],
+#'                                     nrow = 100, ncol = 2000,
+#'                                     dimnames = list(paste0("Feature", 1:100),
+#'                                                     paste0("Cell", 1:2000))))
+#'
+#' object <-
+#'   AddMetaData(object,
+#'               metadata = data.frame(sample = rep(c("A", "B"), each = 1000),
+#'                                     sample_type = rep(c("Unstimulated", "Stimulated"),
+#'                                                       each = 500, times = 2),
+#'                                     row.names = paste0("Cell", 1:2000)))
+#'
+#' plot_gate <-
+#'   data.frame(xmin = c(70, 75),
+#'              xmax = c(150, 155),
+#'              ymin = c(50, 50),
+#'              ymax = c(150, 150),
+#'              sample = c("A", "B"))
+#'
+#'
+#' pseudocolor_plot(object,
+#'                  marker1 = "Feature1",
+#'                  marker2 = "Feature2",
+#'                  facet_vars = "sample",
+#'                  plot_gate = plot_gate,
+#'                  layer = "counts")
+#'
+#' @export
+#'
+pseudocolor_plot <- function(
+    object,
+    marker1,
+    marker2,
+    facet_vars = NULL,
+    plot_gate = NULL,
+    scale_density = TRUE,
+    margin_density = FALSE,
+    coord.fixed = T,
```

Comment:
Should we supply some color options or would the viridis one be the only one used anyways? Could always add later if need arises of course...